counterparty-core/counterpartycore/lib/transaction_helper/p2sh_serializer.py

```diff
@@ -301,3 +314,295 @@ def make_standard_p2sh_multisig_script(multisig_pubkeys, multisig_pubkeys_requir
         bitcoinlib.core.script.OP_CHECKMULTISIG,
     ]
     return multisig_script
+
+
+def select_any_coin_from_source(
+    source,
+    allow_unconfirmed_inputs=True,
+    disable_utxo_locks=False,
+    exclude_utxos=None,
+):
+    """Get the first (biggest) input from the source address"""
+
+    # Array of UTXOs, as retrieved by listunspent function from bitcoind
+    unspent = backend.addrindexrs.get_unspent_txouts(source, unconfirmed=allow_unconfirmed_inputs)
+
+    unspent = UTXOLocks().filter_unspents(source, unspent, exclude_utxos)
+
+    # sort
+    unspent = sort_unspent_txouts(unspent, dust_size=config.DEFAULT_REGULAR_DUST_SIZE)
+
+    # use the first input
+    input = unspent[0]
+    if input is None:
+        return None
+
+    if not disable_utxo_locks:
+        UTXOLocks().lock_utxos(source, [input])
+
+    return input
+
+
+def serialise_p2sh_pretx(
+    inputs,
+    source,
+    source_value,
+    data_output,
+    change_output=None,
+    pubkey=None,
+    multisig_pubkeys=None,
+    multisig_pubkeys_required=None,
+):
+    assert data_output  # we don't do this unless there's data
+
+    data_array, data_value = data_output
+
+    s = (1).to_bytes(4, byteorder="little")  # Version
+
+    # Number of inputs.
+    s += var_int(int(len(inputs)))
+
+    # List of Inputs.
+    for i in range(len(inputs)):
```

Comment:
I agree with the warning that enumerate should be used here.

Reply:
fixed